File: scripts/i18n_checker.js
import * as fs from 'fs';
import * as path from 'path';

// Function to recursively find JSON files
function findJsonFiles(baseDirectory) {
    let jsonFiles = [];
    const files = fs.readdirSync(baseDirectory);
    files.forEach(file => {
        const fullPath = path.join(baseDirectory, file);
        const stat = fs.statSync(fullPath);
        if (stat.isDirectory()) {
            jsonFiles = jsonFiles.concat(findJsonFiles(fullPath));
        } else if (file.endsWith('.json')) {
            jsonFiles.push(fullPath);
        }
    });
    return jsonFiles;
}

// Function to load a JSON file and return its content along with the file name
function loadJson(filePath) {
    return [JSON.parse(fs.readFileSync(filePath, 'utf-8')), path.basename(filePath)];
}

// Function to load keys from all English JSON files
function loadEnKeys(enJsonFiles) {
    let allKeys = new Set();
    enJsonFiles.forEach(jsonFile => {
        const data = JSON.parse(fs.readFileSync(jsonFile, 'utf-8'));
        Object.keys(data).forEach(key => allKeys.add(key));
    });
    return allKeys;
}

// Function to compare the English keys with other locale keys
function compareKeys(enData, otherLocaleData) {
    const enKeys = new Set(Object.keys(enData));
    const otherKeys = new Set(Object.keys(otherLocaleData));

    const missingKeys = [...enKeys].filter(key => !otherKeys.has(key));
    const extraneousKeys = [...otherKeys].filter(key => !enKeys.has(key));

    return [missingKeys, extraneousKeys];
}

// Function to search for keys in files recursively under the search path
function searchForKeyInFiles(key, searchPath) {
    const files = findJsonFiles(searchPath);
    for (const file of files) {
        const contents = fs.readFileSync(file, 'utf-8');
        if (contents.includes(key)) {
            return true; // Key is found
        }
    }
    return false; // Key not found
}

// Function to write a CSV for English labels
function writeEnglishLabelsCsv(enData, outputFile) {
    const headers = `"label_key","value","json_file"\n`;
    let rows = '';
    Object.entries(enData).forEach(([label_key, value]) => {
        rows += `"${label_key}","${value.value}","${value.json_file}"\n`;
    });
    fs.writeFileSync(outputFile, headers + rows);
}

// Function to write a consolidated comparison CSV for all locales
function writeConsolidatedComparisonCsv(comparisonData, outputFile) {
    const headers = `"locale","status","label_key","json_file"\n`;
    let rows = '';
    Object.entries(comparisonData).forEach(([locale, data]) => {
        data.missingKeys.forEach(({ key, jsonFile }) => {
            rows += `"${locale}","missing","${key}","${jsonFile}"\n`;
        });
        data.extraneousKeys.forEach(({ key, jsonFile }) => {
            rows += `"${locale}","extraneous","${key}","${jsonFile}"\n`;
        });
    });
    fs.writeFileSync(outputFile, headers + rows);
}

// Function to write unused keys to a CSV file
function writeUnusedKeysToCsv(unusedKeys, outputFile) {
    const headers = `"unused_key"\n`;
    let rows = '';
    unusedKeys.forEach(key => {
        rows += `"${key}"\n`;
    });
    fs.writeFileSync(outputFile, headers + rows);
}

// Unused keys function
function findUnusedKeys(enLocalePath, searchBasePath, outputDir) {
    const enJsonFiles = findJsonFiles(enLocalePath);
    const allEnKeys = loadEnKeys(enJsonFiles);

    const unusedKeys = [];
    allEnKeys.forEach(key => {
        console.log(`Searching for key: ${key}`);
        if (!searchForKeyInFiles(key, searchBasePath)) {
            unusedKeys.push(key);
        }
    });

    if (outputDir) {
        const outputFile = path.join(outputDir, 'unused_keys.csv');
        writeUnusedKeysToCsv(unusedKeys, outputFile);
        console.log(`Unused keys written to ${outputFile}`);
    } else {
        console.log('Unused keys:');
        console.log(unusedKeys);
    }
}

// Key comparison function
function compareKeysInLocales(basePath, enPath, outputDir) {
    const enFiles = findJsonFiles(enPath);
    const allEnData = {};
    enFiles.forEach(enFile => {
        const [enData, jsonFile] = loadJson(enFile);
        Object.entries(enData).forEach(([key, value]) => {
            allEnData[key] = { value, json_file: jsonFile };
        });
    });

    if (outputDir) {
    const englishLabelsOutputFile = path.join(outputDir, 'english_labels.csv');
    writeEnglishLabelsCsv(allEnData, englishLabelsOutputFile);
    }

    const comparisonData = {};
    const otherLocales = fs.readdirSync(basePath).filter(
        d => fs.statSync(path.join(basePath, d)).isDirectory() && d !== 'en'
    );

    otherLocales.forEach(locale => {
        const localePath = path.join(basePath, locale);
        const localeFiles = findJsonFiles(localePath);
        const localeData = {};
        localeFiles.forEach(localeFile => {
            const [localeJsonData, jsonFile] = loadJson(localeFile);
            Object.entries(localeJsonData).forEach(([key, value]) => {
                localeData[key] = { value, json_file: jsonFile };
            });
        });

        const enKeys = new Set(Object.keys(allEnData));
        const localeKeys = new Set(Object.keys(localeData));
        const missingKeys = [...enKeys].filter(key => !localeKeys.has(key));
        const extraneousKeys = [...localeKeys].filter(key => !enKeys.has(key));

        comparisonData[locale] = {
            missingKeys: missingKeys.map(key => ({ key, jsonFile: allEnData[key].json_file })),
            extraneousKeys: extraneousKeys.map(key => ({ key, jsonFile: localeData[key].json_file }))
        };
    });

    if (outputDir) {
        const consolidatedOutputFile = path.join(outputDir, 'locale_key_comparison_consolidated.csv');
        writeConsolidatedComparisonCsv(comparisonData, consolidatedOutputFile);
    } else {
        console.log('Comparison data:');
        console.log(JSON.stringify(comparisonData, null, 2));
    }

    console.log(`Comparison CSV written to ${consolidatedOutputFile}`);
}

// Function to parse the arguments into key-value pairs
function parseArguments() {
    const args = process.argv.slice(2);
    const argMap = {};
    for (let i = 0; i < args.length; i += 2) {
        const key = args[i];
        const value = args[i + 1];
        if (key.startsWith('--')) {
            argMap[key.slice(2)] = value;
        }
    }
    return argMap;
}

// Main function to parse arguments and invoke appropriate functions
function main() {
    const args = parseArguments();
    const enLocalePath = args['en-locale-path'] || 'public/locales/en';
    const basePath = args['base-path'] || 'public/locales';
    const outputDir = args['output-dir'];
    const searchPath = args['search-path'] || 'src';


    compareKeysInLocales(basePath, enLocalePath, outputDir);
    findUnusedKeys(enLocalePath, searchPath, outputDir);

}

// Run the script
main();

File: scripts/i18n_checker.py

# https://gist.github.com/metalaureate/b4bc4f11c1f47a8100e319884c4edd70

import os
import json
import csv
import argparse

# Function to recursively find JSON files
def find_json_files(base_directory):
    json_files = []
    for root, _, files in os.walk(base_directory):
        for file in files:
            if file.endswith(".json"):
                json_files.append(os.path.join(root, file))
    return json_files

# Function to load a JSON file and return its content along with the file name
def load_json(file_path):
    with open(file_path, 'r', encoding='utf-8') as f:
        return json.load(f), os.path.basename(file_path)

# Function to load keys from all English JSON files
def load_en_keys(en_json_files):
    all_keys = set()
    for json_file in en_json_files:
        with open(json_file, 'r', encoding='utf-8') as f:
            data = json.load(f)
            all_keys.update(data.keys())
    return all_keys

# Function to compare the English keys with other locale keys
def compare_keys(en_data, other_locale_data):
    en_keys = set(en_data.keys())
    other_keys = set(other_locale_data.keys())

    missing_keys = en_keys - other_keys
    extraneous_keys = other_keys - en_keys

    return missing_keys, extraneous_keys

# Function to search for keys in all files recursively under the search path
def search_for_key_in_files(key, search_path):
    for root, _, files in os.walk(search_path):
        for file in files:
            file_path = os.path.join(root, file)
            try:
                with open(file_path, 'r', encoding='utf-8', errors='ignore') as f:
                    contents = f.read()
                    if key in contents:
                        return True  # Key is found
            except:
                pass  # Skip unreadable files
    return False  # Key not found

# Function to write a CSV for English labels
def write_english_labels_csv(en_data, output_file):
    with open(output_file, mode='w', newline='', encoding='utf-8') as file:
        writer = csv.writer(file, quoting=csv.QUOTE_ALL)
        writer.writerow(["label_key", "value", "json_file"])
        for (label_key, json_file), value in en_data.items():
            writer.writerow([label_key, value, json_file])

# Function to write a consolidated comparison CSV for all locales
def write_consolidated_comparison_csv(comparison_data, output_file):
    with open(output_file, mode='w', newline='', encoding='utf-8') as file:
        writer = csv.writer(file, quoting=csv.QUOTE_ALL)
        writer.writerow(["locale", "status", "label_key", "json_file"])
        for locale, data in comparison_data.items():
            for key, json_file in data['missing_keys']:
                writer.writerow([locale, "missing", key, json_file])
            for key, json_file in data['extraneous_keys']:
                writer.writerow([locale, "extraneous", key, json_file])

# Function to write unused keys to a CSV file
def write_unused_keys_to_csv(unused_keys, output_file):
    with open(output_file, mode='w', newline='', encoding='utf-8') as file:
        writer = csv.writer(file, quoting=csv.QUOTE_ALL)
        writer.writerow(["unused_key"])
        for key in unused_keys:
            writer.writerow([key])

# Unused keys function
def find_unused_keys(en_locale_path, search_base_path, output_dir):
    en_json_files = find_json_files(en_locale_path)
    all_en_keys = load_en_keys(en_json_files)
    
    unused_keys = []
    for key in all_en_keys:
        print(f"Searching for key: {key}")
        if not search_for_key_in_files(key, search_base_path):
            unused_keys.append(key)

    output_file = os.path.join(output_dir, 'unused_keys.csv')
    write_unused_keys_to_csv(unused_keys, output_file)
    print(f"Unused keys written to {output_file}")

# Key comparison function
def compare_keys_in_locales(base_path, en_path, output_dir):
    en_files = find_json_files(en_path)
    all_en_data = {}
    for en_file in en_files:
        en_data, json_file = load_json(en_file)
        all_en_data.update({(key, json_file): value for key, value in en_data.items()})
    
    english_labels_output_file = os.path.join(output_dir, 'english_labels.csv')
    write_english_labels_csv(all_en_data, english_labels_output_file)

    comparison_data = {}
    other_locales = [d for d in os.listdir(base_path) if os.path.isdir(os.path.join(base_path, d)) and d != 'en']

    for locale in other_locales:
        locale_path = os.path.join(base_path, locale)
        locale_files = find_json_files(locale_path)
        locale_data = {}
        for locale_file in locale_files:
            locale_json_data, json_file = load_json(locale_file)
            locale_data.update({(key, json_file): value for key, value in locale_json_data.items()})
        
        en_keys = set(key for key, _ in all_en_data.keys())
        locale_keys = set(key for key, _ in locale_data.keys())
        missing_keys = en_keys - locale_keys
        extraneous_keys = locale_keys - en_keys

        missing_keys_with_file = [(key, file) for (key, file) in all_en_data.keys() if key in missing_keys]
        extraneous_keys_with_file = [(key, file) for (key, file) in locale_data.keys() if key in extraneous_keys]

        comparison_data[locale] = {
            'missing_keys': missing_keys_with_file,
            'extraneous_keys': extraneous_keys_with_file
        }

    consolidated_output_file = os.path.join(output_dir, 'locale_key_comparison_consolidated.csv')
    write_consolidated_comparison_csv(comparison_data, consolidated_output_file)
    print(f"Comparison CSV written to {consolidated_output_file}")

# Main function to parse arguments and invoke appropriate functions
def main():
    parser = argparse.ArgumentParser(description="CLI tool for comparing locale JSON keys and finding unused keys.")
    parser.add_argument("mode", choices=["compare", "unused"], help="Mode of operation: 'compare' or 'unused'.")
    parser.add_argument("--en-locale-path", required=True, help="Path to the English locale directory.")
    parser.add_argument("--base-path", required=True, help="Base path for all locales (for comparison).")
    parser.add_argument("--output-dir", required=True, help="Directory to store the output CSV files.")
    parser.add_argument("--search-path", required=False, help="Path to search for unused keys (required for 'unused' mode).")

    args = parser.parse_args()

    if args.mode == "compare":
        compare_keys_in_locales(args.base_path, args.en_locale_path, args.output_dir)
    elif args.mode == "unused":
        if not args.search_path:
            print("Error: --search-path is required for 'unused' mode.")
            return
        find_unused_keys(args.en_locale_path, args.search_path, args.output_dir)

if __name__ == "__main__":
    main()

# python tool.py compare --en-locale-path /path/to/en --base-path /path/to/locales --output-dir /path/to/output
# python tool.py unused --en-locale-path /path/to/en --base-path /path/to/locales --search-path /path/to/src --output-dir /path/to/output
  
# Example usage:
# python i18n_checker.py compare --en-locale-path /Users/possum/Projects/tari/universe/public/locales/en --base-path /Users/possum/Projects/tari/universe/public/locales --output-dir ~/Downloads
# python i18n_checker.py unused --en-locale-path /Users/possum/Projects/tari/universe/public/locales/en --base-path /Users/possum/Projects/tari/universe/public/locales --search-path /Users/possum/Projects/tari/universe/src --output-dir ~/Downloads
